A98-debug-protocol.md

```diff
@@ -0,0 +1,224 @@
+    repeated Entity entities = 1;
+}
+```
+
+**GetEntity** allows polling a single entity's state.
+
+```
+message GetEntityRequest {
+    // The identifier of the entity to get.
+    int64 id = 1;
+}
+
+message GetEntityResponse {
+    // The Entity that corresponds to the requested id.  This field
+    // should be set.
+    Entity entity = 1;
+}
+```
+
+Finally, **QueryTrace** allows rich queries of live trace data to be pulled from an instance.
+
+```
+message QueryTraceRequest {
+    // The identifier of the entity to query.
+    int64 id = 1;
+    // The name of the trace to query.
+    string name = 2;
+    // The amount of time to query. Implementations may arbitrarily cap this.
+    google.protobuf.Duration duration = 3;
+    // Implementation defined query arguments.
+    repeated google.protobuf.Any args = 4;
+}
+
+message QueryTraceResponse {
+    // The events in the trace.
+    repeated TraceEvent events = 1;
+    // Number of events matched by the trace.
+    // This may be higher than the number returned if memory limits were exceeded.
+    int64 num_events_matched = 2;
+}
+```
+
+The idea here is that these traces will be very detailed - beyond what can feasibly be stored in the historical traces.
+So instead of needing to store this data in historical traces *in case* it is needed, we query and collect this data on demand for small windows of time.
+
+### Well Known Data
+
+A separate `well_known_data.proto` file will be maintained, with debug state that's common across implementations described within it.
+
+The initial state will be:
+
+```
+// Channel connectivity state - attached to kind "channel" and "subchannel".
+// These come from the specified states in this document:
+// https://github.com/grpc/grpc/blob/master/doc/connectivity-semantics-and-api.md
+message ChannelConnectivityState {
+    enum State {
+        UNKNOWN = 0;
+        IDLE = 1;
+        CONNECTING = 2;
+        READY = 3;
+        TRANSIENT_FAILURE = 4;
+        SHUTDOWN = 5;
+    }
+    State state = 1;
+}
+
+// Channel target information. Attached to kind "channel" and "subchannel".
+message ChannelTarget {
+    // The target this channel originally tried to connect to.  May be absent
+    string target = 1;
+}
+```
+
+## Rationale
+
+The new interface is heavily inspired by channelz, but removes rigid node associations and predefined data sets.
+Much of the infrastructure required for channelz can be shared with the debug interface, and indeed one can be implemented atop the other with good fidelity (in either direction).
+
+Node ids are consistent between the protocols so that one can take a node id from the debug interface and query channelz data directly (and vice versa).
+This will allow consumers of the protocols to gradually transition from one to the other.
+It also allows re-using the book-keeping already in existence for channelz in implementations, without adding another kind of book-keeping to the mix.
+
+Why not just extend channelz?
```

Comment:
We could potentially add the new types and methods to the existing channelz interface if we wanted.

One thing worth noting is that channelz was designed for the use case of having a common UI that is able to show data for all languages with a common shape of tree.  If we could find a way to support both use-cases in a common API, that seems ideal, but we'd have to brainstorm whether there are viable ways to do that.

Reply:
ack. I think we've discussed offline directions.